### PR TITLE
FEATURE: TMA-783 Support authenticate to ADS instance using SST

### PR DESCRIPTION
**Pull request by phong-nguyen-duy**

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by phong-nguyen-duy** [comment]
ok to test

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: lib/gooddata_datawarehouse/datawarehouse.rb [hunk 1]
```diff
@@ -192,11 +190,12 @@ def execute_select(sql, opts={})
     end
 
     def connect
-      Sequel.connect @jdbc_url,
-        :username => @username,
-        :password => @password do |connection|
-          yield(connection)
+      if @username.nil? || @password.nil?
```
Comment (ostumpf) [inline lib/gooddata_datawarehouse/datawarehouse.rb]: Do we want to address empty strings somehow? Maybe `@username.to_s.nil? || @password.to_s.nil?`

File: lib/gooddata_datawarehouse/datawarehouse.rb [hunk 2]
```diff
@@ -192,11 +190,12 @@ def execute_select(sql, opts={})
...
+        connection = Sequel.connect(@jdbc_url, :driver => Java.com.gooddata.dss.jdbc.driver.DssDriver, :jdbc_properties => {'sst' => @sst_token})
+      else
+        connection = Sequel.connect(@jdbc_url, :username => @username, :password => @password)
       end
+      yield(connection)
```
Comment (ostumpf) [inline lib/gooddata_datawarehouse/datawarehouse.rb]: I think this is wrong. The old implementation passed a block to the `connect` method, and therefore `Sequel` closed the session once the block (containing `yield`) finished. The new implementation on the other hand never closes the session.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: lib/gooddata_datawarehouse/datawarehouse.rb (hunk 1 above)
Comment (phong-nguyen-duy) [reply lib/gooddata_datawarehouse/datawarehouse.rb]: Yes, I used @username.to_s.empty? to check nil or empty string.

File: lib/gooddata_datawarehouse/datawarehouse.rb (hunk 2 above)
Comment (phong-nguyen-duy) [reply lib/gooddata_datawarehouse/datawarehouse.rb]: Nice catch. I fixed it by closing connection in ensure block.

File: lib/gooddata_datawarehouse/datawarehouse.rb [hunk 3]
```diff
@@ -192,11 +190,14 @@ def execute_select(sql, opts={})
     end
 
     def connect
-      Sequel.connect @jdbc_url,
-        :username => @username,
-        :password => @password do |connection|
-          yield(connection)
+      if @username.to_s.empty? || @password.to_s.empty?
+        @connection = Sequel.connect(@jdbc_url, :driver => Java.com.gooddata.dss.jdbc.driver.DssDriver, :jdbc_properties => {'sst' => @sst_token})
```
Comment (ostumpf) [inline lib/gooddata_datawarehouse/datawarehouse.rb]: Why do we actually need the `:driver => Java.com.gooddata.dss.jdbc.driver.DssDriver` param? It's not present in the `else` branch.

Comment (phong-nguyen-duy) [reply lib/gooddata_datawarehouse/datawarehouse.rb]: It is mentioned in this document https://confluence.intgdc.com/display/~petr.tichy/ETL+pipeline+using+SST+and+impersonalization

Comment (ostumpf) [reply lib/gooddata_datawarehouse/datawarehouse.rb]: ok I see

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata_datawarehouse-opensource-repository-check/1/ : SUCCESS in 9s

**Comment by vu-pham** [comment]
@jimirocks, @juras-gdc, @jakub-plichta Do you have the permission to release gooddata_warehouse gem? If that, please release it. Many thanks.

**Pull request merged**